Race in `set_qdr_connection_info_details` and `qdr_connection_insert_column_CT` in system_tests_http2_tls

https://github.com/skupperproject/skupper-router/runs/6213145255?check_suite_focus=true#step:26:1626

```
E           WARNING: ThreadSanitizer: data race (pid=3388)
E             Write of size 1 at 0x7b200000a550 by main thread:
E               #0 set_qdr_connection_info_details /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2646 (skrouterd+0x4e0c88)
E               #1 handle_incoming_tls /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2725 (skrouterd+0x4e0c88)
E               #2 handle_incoming_http /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2792 (skrouterd+0x4e1102)
E               #3 handle_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:3267 (skrouterd+0x4e19d3)
E               #4 handle_event_with_context /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:780 (skrouterd+0x4c91b9)
E               #5 do_handle_raw_connection_event /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:786 (skrouterd+0x4c91b9)
E               #6 handle /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1063 (skrouterd+0x4c91b9)
E               #7 thread_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1095 (skrouterd+0x4caf17)
E               #8 qd_server_run /home/runner/work/skupper-router/skupper-router/skupper-router/src/server.c:1491 (skrouterd+0x4cba8c)
E               #9 main_process /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:105 (skrouterd+0x42711c)
E               #10 main /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:359 (skrouterd+0x42651c)
E           
E             Previous read of size 1 at 0x7b200000a550 by thread T2:
E               #0 qdr_connection_insert_column_CT /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/agent_connection.c:215 (skrouterd+0x48be35)
E               #1 qdr_agent_write_connection_CT /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/agent_connection.c:315 (skrouterd+0x48c4a1)
E               #2 qdra_connection_get_first_CT /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/agent_connection.c:364 (skrouterd+0x48c4a1)
E               #3 qdrh_query_get_first_CT /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/agent.c:517 (skrouterd+0x48722d)
E               #4 router_core_thread /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/router_core_thread.c:247 (skrouterd+0x4ab50a)
E               #5 _thread_init /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:172 (skrouterd+0x47d99d)
E           
E             Location is heap block of size 128 at 0x7b200000a500 allocated by main thread:
E               #0 posix_memalign <null> (libtsan.so.0+0x32a23)
E               #1 qd_alloc /home/runner/work/skupper-router/skupper-router/skupper-router/src/alloc_pool.c:409 (skrouterd+0x44bde9)
E               #2 new_qdr_connection_info_t /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/router_core.c:41 (skrouterd+0x4a587c)
E               #3 qdr_connection_info /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/connections.c:177 (skrouterd+0x48ebdc)
E               #4 qdr_http_connection_egress /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:3113 (skrouterd+0x4e2bea)
E               #5 qd_http2_configure_connector /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:3411 (skrouterd+0x4e349c)
E               #6 qd_dispatch_configure_http_connector /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http_common.c:191 (skrouterd+0x42a5f6)
E               #7 ffi_call_unix64 <null> (libffi.so.6+0x6c03)
E               #8 main_process /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:92 (skrouterd+0x4270bc)
E               #9 main /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:359 (skrouterd+0x42651c)
E           
E             Thread T2 (tid=3391, running) created by main thread at:
E               #0 pthread_create <null> (libtsan.so.0+0x5bef5)
E               #1 sys_thread /home/runner/work/skupper-router/skupper-router/skupper-router/src/posix/threading.c:181 (skrouterd+0x47de2c)
E               #2 qdr_core /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/router_core.c:122 (skrouterd+0x4a5dd1)
E               #3 qd_router_setup_late /home/runner/work/skupper-router/skupper-router/skupper-router/src/router_node.c:2058 (skrouterd+0x4c44f8)
E               #4 ffi_call_unix64 <null> (libffi.so.6+0x6c03)
E               #5 main_process /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:92 (skrouterd+0x4270bc)
E               #6 main /home/runner/work/skupper-router/skupper-router/skupper-router/router/src/main.c:359 (skrouterd+0x42651c)
E           
E           SUMMARY: ThreadSanitizer: data race /home/runner/work/skupper-router/skupper-router/skupper-router/src/adaptors/http2/http2_adaptor.c:2646 in set_qdr_connection_info_details
```

--- a/src/adaptors/http2/http2_adaptor.c
+++ b/src/adaptors/http2/http2_adaptor.c
@@ -2634,28 +2634,45 @@ static void handle_outgoing_tls(qdr_http2_connection_t *conn, const pn_raw_buffe
 
 static void set_qdr_connection_info_details(qdr_http2_connection_t *conn)
 {
-    qdr_connection_info_t *conn_info = conn->qdr_conn->connection_info;
-    if (conn_info) {
-        const char *protocol_info;
-        size_t protocol_info_length;
-        if (pn_tls_get_protocol_version(conn->tls_session, &protocol_info, &protocol_info_length)) {
-            char *protocol_ver = qd_calloc(protocol_info_length + 1, sizeof(char));
-            memmove(protocol_ver, protocol_info, protocol_info_length);
-            free(conn_info->ssl_proto);
-            conn_info->ssl_proto = protocol_ver;
-            conn_info->ssl = true;
-            conn_info->is_encrypted = true;
-        }
-        if (pn_tls_get_cipher(conn->tls_session, &protocol_info, &protocol_info_length)) {
-            char *protocol_cipher = qd_calloc(protocol_info_length + 1, sizeof(char));
-            memmove(protocol_cipher, protocol_info, protocol_info_length);
-            free(conn_info->ssl_cipher);
-            conn_info->ssl_cipher = protocol_cipher;
-        }
-        if (conn->config->authenticate_peer) {
-            conn_info->is_authenticated = true;
-        }
+    const char *protocol_info;
+    size_t protocol_info_length;
+    char *protocol_ver = 0;
+    char *protocol_cipher = 0;
+    //
+    // Ask the Proton TLS API for protocol version and protol cipher.
+    //
+    if (pn_tls_get_protocol_version(conn->tls_session, &protocol_info, &protocol_info_length)) {
+        protocol_ver = qd_calloc(protocol_info_length + 1, sizeof(char));
+        memmove(protocol_ver, protocol_info, protocol_info_length);
     }
+    if (pn_tls_get_cipher(conn->tls_session, &protocol_info, &protocol_info_length)) {
+        protocol_cipher = qd_calloc(protocol_info_length + 1, sizeof(char));
+        memmove(protocol_cipher, protocol_info, protocol_info_length);
+    }
+
+    qdr_connection_info_t *conn_info = conn->qdr_conn->connection_info;
+
+    //
+    // Lock using the connection_info_lock before setting the values on the
+    // connection_info. This same lock is being used in the agent_connection.c's qdr_connection_insert_column_CT
+    //
+    sys_mutex_lock(conn_info->connection_info_lock);
+    free(conn_info->ssl_cipher);
+    conn_info->ssl_cipher = 0;
+    free(conn_info->ssl_proto);
+    conn_info->ssl_proto = 0;
+    conn_info->ssl = true;
+    conn_info->is_encrypted = true;
+    if (protocol_cipher) {
+        conn_info->ssl_cipher = protocol_cipher;
+    }
+    if (protocol_ver) {
+        conn_info->ssl_proto = protocol_ver;
+    }
+    if (conn->config->authenticate_peer) {
+        conn_info->is_authenticated = true;
+    }
+    sys_mutex_unlock(conn_info->connection_info_lock);
 }
 
 /**

--- a/src/router_core/agent_connection.c
+++ b/src/router_core/agent_connection.c
@@ -130,6 +130,8 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
     if (as_map)
         qd_compose_insert_string(body, qdr_connection_columns[col]);
 
+    sys_mutex_lock(conn->connection_info->connection_info_lock);
+
     switch(col) {
     case QDR_CONNECTION_NAME:
         qd_compose_insert_string2(body, "connection/", conn->connection_info->host);
@@ -175,9 +177,7 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_IS_AUTHENTICATED:
-        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->is_authenticated);
-        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_USER:
@@ -188,27 +188,21 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_IS_ENCRYPTED:
-        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->is_encrypted);
-        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLPROTO:
-        sys_mutex_lock(conn->connection_info->connection_info_lock);
         if (conn->connection_info->ssl_proto[0] != '\0')
             qd_compose_insert_string(body, conn->connection_info->ssl_proto);
         else
             qd_compose_insert_null(body);
-        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLCIPHER:
-        sys_mutex_lock(conn->connection_info->connection_info_lock);
         if (conn->connection_info->ssl_cipher[0] != '\0')
             qd_compose_insert_string(body, conn->connection_info->ssl_cipher);
         else
             qd_compose_insert_null(body);
-        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLSSF:
@@ -220,9 +214,7 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_SSL:
-        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->ssl);
-        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_OPENED:
@@ -310,6 +302,8 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
     }
     break;
     }
+
+    sys_mutex_unlock(conn->connection_info->connection_info_lock);
 }
 
 

--- a/src/router_core/agent_connection.c
+++ b/src/router_core/agent_connection.c
@@ -175,7 +175,9 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_IS_AUTHENTICATED:
+        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->is_authenticated);
+        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_USER:
@@ -186,21 +188,27 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_IS_ENCRYPTED:
+        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->is_encrypted);
+        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLPROTO:
+        sys_mutex_lock(conn->connection_info->connection_info_lock);
         if (conn->connection_info->ssl_proto[0] != '\0')
             qd_compose_insert_string(body, conn->connection_info->ssl_proto);
         else
             qd_compose_insert_null(body);
+        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLCIPHER:
+        sys_mutex_lock(conn->connection_info->connection_info_lock);
         if (conn->connection_info->ssl_cipher[0] != '\0')
             qd_compose_insert_string(body, conn->connection_info->ssl_cipher);
         else
             qd_compose_insert_null(body);
+        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_SSLSSF:
@@ -212,7 +220,9 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
         break;
 
     case QDR_CONNECTION_SSL:
+        sys_mutex_lock(conn->connection_info->connection_info_lock);
         qd_compose_insert_bool(body, conn->connection_info->ssl);
+        sys_mutex_unlock(conn->connection_info->connection_info_lock);
         break;
 
     case QDR_CONNECTION_OPENED:

--- a/src/router_core/connections.c
+++ b/src/router_core/connections.c
@@ -204,6 +204,7 @@ qdr_connection_info_t *qdr_connection_info(bool             is_encrypted,
     connection_info->ssl_ssf = ssl_ssf;
     connection_info->ssl     = ssl;
     connection_info->streaming_links = streaming_links;
+    connection_info->connection_info_lock = sys_mutex();
     return connection_info;
 }
 
@@ -217,6 +218,7 @@ static void qdr_connection_info_free(qdr_connection_info_t *ci)
     free(ci->ssl_cipher);
     free(ci->user);
     free(ci->version);
+    sys_mutex_free(ci->connection_info_lock);
     pn_data_free(ci->connection_properties);
     free_qdr_connection_info_t(ci);
 }

--- a/src/router_core/router_core_private.h
+++ b/src/router_core/router_core_private.h
@@ -625,6 +625,7 @@ struct qdr_connection_info_t {
     bool                        ssl;
     int                         ssl_ssf; //ssl strength factor
     char                       *version; // if role is router or edge
+    sys_mutex_t                *connection_info_lock;
 };
 
 ALLOC_DECLARE(qdr_connection_info_t);

